### PR TITLE
p-ata: Implement bump hints

### PR DESCRIPTION
**Pull request by grod220**
Makes use of the `bump_hint` argument in the processor fn. When a hint is supplied, it is verified by checking that no higher bump produces an off-curve PDA rather than running `find_program_address`'s full descending search.

The savings is exactly one `is_on_curve()` call (~130 CU). Though take note of the idempotent-account-exists benchmark comment.

### CODE REVIEW

File: pinocchio/program/benches/compute_units.md [hunk 1]
```diff
@@ -1,3 +1,30 @@
...
+| create (token-2022) | 5246 | +1 |
+| create_with_args (token-2022) | 5051 | -133 |
+| create_idempotent (new, spl-token) | 4300 | +2 |
+| create_with_args_idempotent (new, spl-token) | 4106 | -131 |
+| create_idempotent (new, token-2022) | 5613 | +1 |
+| create_with_args_idempotent (new, token-2022) | 5420 | -130 |
+| create_idempotent (existing, spl-token) | 548 | -- |
+| create_with_args_idempotent (existing, spl-token) | 605 | +38 |
+| create_idempotent (existing, token-2022) | 1634 | -- |
+| create_with_args_idempotent (existing, token-2022) | 1694 | +41 |
```
Comment (grod220) [inline pinocchio/program/benches/compute_units.md:18]: This is a cost caused by the [extra check](https://github.com/solana-program/associated-token-account/pull/250/changes#diff-33f75be8b0dcfac2751512480f772dac68510351487f75361baa803242e34e2cR85-R91) required for security reasons in the idempotent case where the account already exists.

It's a bit of a trade off the client will need to weigh the probabilities of. Options:
- Skip bump, account exists baseline
- Skip bump, account doesn't exist baseline
- Pass bump, account exists: +38 CU
- Pass bump, account doesn't exist: −131 CU

Comment (joncinque) [reply pinocchio/program/benches/compute_units.md:18]: This is probably reasonable. Most likely off-chain users will end up hitting the bump + account exists case more often, since an on-chain program can just check the account and decide to not perform a CPI at all.

**Review by joncinque — DISMISSED** [review]
Looks great!

File: pinocchio/interface/src/pda.rs [hunk 2]
```diff
@@ -1,6 +1,6 @@
 //! Address derivation helpers for Associated Token Account program-derived addresses.
 
-use solana_address::Address;
+use {pinocchio::error::ProgramError, solana_address::Address};
```
Comment (febo) [inline pinocchio/interface/src/pda.rs]: nit: You can use the re-exports from pinocchio.
```suggestion
use pinocchio::{Address, error::ProgramError};
```

File: pinocchio/program/src/create.rs [hunk 3]
```diff
@@ -71,6 +82,13 @@ pub(crate) fn process_create_associated_token_account(
                     if token_account.mint() != mint.address() {
                         return Err(ProgramError::InvalidAccountData);
                     }
+                    // `derive_address_with_bump_hint()` rejects higher off-curve bumps but
+                    // doesn't check whether the hinted bump itself is off-curve. Create paths
+                    // validate that through `invoke_signed()`, but this no-op path returns early,
+                    // so check it here.
+                    if bump_hint.is_some() && derived_ata_addr.is_on_curve() {
```
Comment (febo) [inline pinocchio/program/src/create.rs:89]: Do we need this check? If the account exists + has data, that means it is off-curve, no? Like, it would not have been created if it wasn't off-curve.

File: pinocchio/interface/src/pda.rs [hunk 4]
```diff
@@ -54,4 +54,37 @@ impl AssociatedTokenPda {
...
+
+        if let Some(first_higher_bump) = bump.checked_add(1) {
+            for higher_bump in first_higher_bump..=u8::MAX {
+                let higher_bump_addr =
+                    Address::derive_address(&seeds, Some(higher_bump), program_id);
+                if !higher_bump_addr.is_on_curve() {
+                    return Err(ProgramError::InvalidSeeds);
+                }
+            }
+        }
```
Comment (febo) [inline pinocchio/interface/src/pda.rs]: nit: I would check whether this is cheaper or not.
```suggestion
        if bump < u8::MAX {
            for higher_bump in (bump + 1)..=u8::MAX {
                let higher_bump_addr =
                    Address::derive_address(&seeds, Some(higher_bump), program_id);
                if !higher_bump_addr.is_on_curve() {
                    return Err(ProgramError::InvalidSeeds);
                }
            }
        }
```
You could even use a "hint" on the if-statement, `unlikely(bump < u8::MAX)`.

**Review by febo — COMMENTED** [review]
Just have a couple of suggestions.

File: pinocchio/interface/src/pda.rs (hunk 4 above)
Comment (grod220) [reply pinocchio/interface/src/pda.rs]: The CUs don't change with `if bump < u8::MAX {`, but I find it more readable will swap. 

But using `unlikely(bump < u8::MAX)` produces these results:
| Case | No Hint CU | With unlikely CU | Diff |
| --- | ---: | ---: | ---: |
| bump=255 | 3013 | 3009 | -4 |
| bump=254 | 3378 | 3393 | +15 |
| bump=253 | 3740 | 3772 | +32 |
| bump=252 | 4102 | 4151 | +49 |

But I'm not sure this is a good idea as it's not that unlikely. From the 100,000-sample validation run:
```
bump 255: 50.179%
bump 254: 25.066%
bump 253: 12.454%
bump 252:  6.225%
bump 251:  3.073%
bump 250:  1.450%
bump <=249: 1.553%
```

Basically in 50% of the cases, it will loop and take a penalty

File: pinocchio/program/src/create.rs (hunk 3 above)
Comment (grod220) [reply pinocchio/program/src/create.rs:89]: But someone can create a token account at an on-curve address right? This could be an attack vector.

Comment (febo) [reply pinocchio/program/src/create.rs:89]: <del>It should not be possible. This is a PDA and the derivation was validated at creation, so if it exists it should be valid (off-curve).</del>

<del>Having another look at this, I think this check needs to be on the  creation only path (not idempotent). That path is not validating the derivation when there is a bump hint.</del>

File: pinocchio/interface/src/pda.rs (hunk 4 above)
Comment (febo) [reply pinocchio/interface/src/pda.rs]: What about using likely? 😀

File: pinocchio/program/src/create.rs (hunk 3 above)
Comment (febo) [reply pinocchio/program/src/create.rs:89]: <del>On a second thought, the runtime validates that a PDA is off-curve, so I don't think we need an explicit check on the create only path.</del>

Comment (febo) [reply pinocchio/program/src/create.rs:89]: Scratch that - the owner of the account is the token program, so you could indeed have an account with a derivation on-curve created outside the ATA program.

Comment (grod220) [reply pinocchio/program/src/create.rs:89]: I think in practice it isn't a feasible attack though. The attacker would need to:
- Find a (wallet, mint, bump) combo where the derived address lands on-curve
- Create the token account at that exact address
- Call `CreateIdempotent` with that bump_hint arg

If the program didn't have this check it would succeed. However, step 2 isn't feasible to grind a keypair for exact addresses like that.

That said, I think it's worth keeping the check to make this edge case invariant impossible. [This test](https://github.com/solana-program/associated-token-account/pull/250/changes#diff-8b2f5ab25b985b2a4bcf3b20a588167790f6fc603b8ed6d17fb44d070672e5b9R176) exercises that guard code path fyi.

File: pinocchio/interface/src/pda.rs (hunk 4 above)
Comment (grod220) [reply pinocchio/interface/src/pda.rs]: No CU impact using `likely()`

File: pinocchio/program/src/create.rs (hunk 3 above)
Comment (febo) [reply pinocchio/program/src/create.rs:89]: Yeah, we should keep the check since while unlikely, it is feasible.

File: pinocchio/interface/src/pda.rs (hunk 4 above)
Comment (febo) [reply pinocchio/interface/src/pda.rs]: Oh well, we tried. 😅

**Review by febo — APPROVED** [review]
